easybuild/easyblocks/o/openblas.py

```diff
@@ -94,3 +168,132 @@ def sanity_check_step(self):
+            if spec[0] == 'march' or spec[0] == '-march':
+                target_arch = spec[1]
+
+        return target_arch.lower()
+
+    def _get_openblas_targets(self, targetfile):
+        """
+        Parse the openblas target file and generate a list of targets.
+        :param targetfile: A file with OpenBLAS targets.
+        """
+        targets = []
+
+        if os.path.isfile(targetfile):
+            # Assumption, the OpenBLAS TargetList.txt has one target per line and that
+            # single words on a line is a target if they match a simple regexp...
+            re_target = re.compile(r'^[A-Z0-9_]+$')
+            for line in read_file(targetfile).splitlines():
+                match = re_target.match(line)
+                if match is not None:
+                    targets.append(line.strip().lower())
+        else:
+            print_warning("Unable to find OpenBLAS targetfile '%s'" % os.path.realpath(targetfile))
+
+        return targets
+
+    def _set_dynamic_architecture(self, default_opts):
+        """
+        Sets the DYNAMIC_ARCH option for OpenBLAS, building a library that chooses
+        an optimized kernel at runtime. Also removes any previous TARGET setting, if any.
+        """
+        default_opts['DYNAMIC_ARCH'] = 1
+        default_opts.pop(TARGET, None)
+        self._dynamic_target = True
+
+    def _get_mapped_target(self, march, openblas_targets):
+        """
+        Attempts to match the given march in the list of openblas targets.
+        If archspec is installed, will try to match directly or follow ancestors for
+        architectures that will work.
+        Returns None if no target was found.
+        """
+
+        result = None
+
+        if HAVE_ARCHSPEC:
+            self.log.info("Using archspec to match optarch to openblas targets.")
+
+            openblas_arch = set(['alpha', 'arm', 'ia64', 'mips', 'mips64',
+                                 'power', 'sparc', 'zarch'])
+            openblas_arch_map = {
+                'amd64': 'x86_64',
+                'powerpc64': 'power',
+                'i386': 'x86',
+                'aarch64': 'arm64',
+            }
+            openblas_arch.update(openblas_arch_map.keys())
+            openblas_arch.update(openblas_arch_map.values())
+
+            skylake = set(["skylake"])
+            available_targets = set(openblas_targets) | skylake | openblas_arch
+
+            try:
+                uarch = archspec_cpu.TARGETS[march]
+            except KeyError:
+                warning_string = "Archspec was asked to find '" + march + "' as an architecture, but failed!"
+                print_warning(warning_string)
+                self.log.warning(warning_string)
+                return None
+
+            if uarch.name in available_targets:
+                result = uarch.name
+            else:
+                self.log.info("No direct match for '" + march + "' between archspec and OpenBLAS, traversing ancestry.")
```

Comment:
Don't we use formatted logs instead of string concatting?

Reply:
Probably? I'll have to look at that.